src/Stratis.Bitcoin.Features.BlockStore/BlockStoreQueue.cs

```diff
@@ -179,60 +179,63 @@ public async Task InitializeAsync()
-                        return Task.FromResult(tx);
-                    }
-                }
-            }
+            Block block = await this.GetBlockAsync(blockId).ConfigureAwait(false);
 
-            return this.blockRepository.GetTransactionByIdAsync(trxid);
+            return block.Transactions.FirstOrDefault(x => x.GetHash() == trxid);
         }
 
         /// <inheritdoc/>
-        public Task<uint256> GetBlockIdByTransactionIdAsync(uint256 trxid)
+        public async Task<uint256> GetBlockIdByTransactionIdAsync(uint256 trxid)
         {
             lock (this.blocksCacheLock)
             {
-                // The result is meaningless unless its selected from the current chain.
-                for (ChainedHeader chainedHeader = this.chain.Tip; chainedHeader != null; chainedHeader = chainedHeader.Previous)
+                // According to SaveBlockAsync the last block in the list is considered to be on the current main chain.
+                if (this.chain.GetBlock(this.BlockStoreCacheTip.HashBlock) != null)
                 {
-                    if (!this.pendingBlocksCache.TryGetValue(chainedHeader.HashBlock, out ChainedHeaderBlock chainedHeaderBlock))
-                        break;
+                    // Indeed so let's proceed.
+                    for (ChainedHeader chainedHeader = this.BlockStoreCacheTip; chainedHeader != null; chainedHeader = chainedHeader.Previous)
+                    {
+                        if (!this.pendingBlocksCache.TryGetValue(chainedHeader.HashBlock, out ChainedHeaderBlock chainedHeaderBlock))
+                            break;
 
-                    bool exists = chainedHeaderBlock.Block.Transactions.Any(x => x.GetHash() == trxid);
+                        bool exists = chainedHeaderBlock.Block.Transactions.Any(x => x.GetHash() == trxid);
 
-                    if (exists)
-                    {
-                        uint256 blockId = chainedHeaderBlock.Block.GetHash();
+                        if (exists)
+                        {
+                            uint256 blockId = chainedHeaderBlock.Block.GetHash();
 
-                        this.logger.LogTrace("Block Id '{0}' with tx '{1}' was found in the pending blocks cache.", blockId, trxid);
-                        return Task.FromResult(blockId);
+                            this.logger.LogTrace("Block Id '{0}' with tx '{1}' was found in the pending blocks cache.", blockId, trxid);
+                            return blockId;
+                        }
                     }
                 }
             }
 
-            return this.blockRepository.GetBlockIdByTransactionIdAsync(trxid);
+            // Now look in the repository.
+            uint256 blockId2 = await this.blockRepository.GetBlockIdByTransactionIdAsync(trxid).ConfigureAwait(false);
+
+            // The repository may be dirty. Check that this is a valid block.
+            if (blockId2 == null || this.chain.GetBlock(blockId2) == null)
```

Comment:
We should not use `chain` this will be deleted soon, instead use CM.Tip or ChainState.Tip

Reply:
I assume the existing code referencing `chain` in this file will also need to be updated making this both a bug fix and an enhancement?

Reply:
Changed to:
```
            // The repository may be dirty. Check that this is a valid block.
            if (blockId2 == null || this.chainState.ConsensusTip.FindAncestorOrSelf(blockId2) == null)
                return null;
```